### PR TITLE
Async writes to pointer

### PR DESCRIPTION
**Pull request by CallanTaylor**

### CODE REVIEW

File: tenets/codelingo/go/writes-to-async-pointer/example.go [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+func newThingPointer(name string) *thing {
+	return &thing{name}
+}
+
+func newThing(name string) thing {
+	return thing{name}
+}
+
+func main() {
```
Comment (fwereade) [inline tenets/codelingo/go/writes-to-async-pointer/example.go]: is there a reason we usually stick all the issues in a single `main` func? One function per issue would be a bit more readable – or at least breaking out the `t2` case to demonstrate that we don't raise an issue when there are no writes happening.

Also, documentation on the components would be really helpful – I look to the examples first to understand the tenet. I'll add some vague suggestions.

File: tenets/codelingo/go/writes-to-async-pointer/example.go [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+func getName(t thing) {
+	fmt.Println(t.name)
+}
+
+func setNameOnPointer(t *thing) {
+	t.name = "car"
+}
+
+func newThingPointer(name string) *thing {
```
Comment (fwereade) [inline tenets/codelingo/go/writes-to-async-pointer/example.go]: I don't think this has any explanatory power over using `&thing{name}` in the source directly

File: tenets/codelingo/go/writes-to-async-pointer/example.go [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+	return &thing{name}
+}
+
+func newThing(name string) thing {
```
Comment (fwereade) [inline tenets/codelingo/go/writes-to-async-pointer/example.go]: ditto newThingPointer

File: tenets/codelingo/go/writes-to-async-pointer/example.go [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+import (
+	"fmt"
+)
+
+type thing struct {
+	name string
+}
+
+func getNameFromPointer(t *thing) {
```
Comment (fwereade) [inline tenets/codelingo/go/writes-to-async-pointer/example.go]: `// readVulnerableField accesses data which may be vulnerable to concurrent writes`?

File: tenets/codelingo/go/writes-to-async-pointer/example.go [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+	fmt.Println(t.name)
+}
+
+func getName(t thing) {
```
Comment (fwereade) [inline tenets/codelingo/go/writes-to-async-pointer/example.go]: `readSafeField` as above?

File: tenets/codelingo/go/writes-to-async-pointer/example.go [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+	fmt.Println(t.name)
+}
+
+func setNameOnPointer(t *thing) {
```
Comment (fwereade) [inline tenets/codelingo/go/writes-to-async-pointer/example.go]: `writeVulnerableField`?

File: tenets/codelingo/go/writes-to-async-pointer/example.go [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+	return thing{name}
+}
+
+func main() {
+
+	t := newThingPointer("Pointer")
+
+	getNameFromPointer(t)
+
+	go getNameFromPointer(t)
```
Comment (fwereade) [inline tenets/codelingo/go/writes-to-async-pointer/example.go]: this block really makes me think we need a sense of read-sections and write-sections

would we catch `go read(); write()`? 'cos it's fundamentally the same issue

Comment (CallanTaylor) [reply tenets/codelingo/go/writes-to-async-pointer/example.go]: Are you referring to the Read and Write methods of the io package?

Comment (fwereade) [reply tenets/codelingo/go/writes-to-async-pointer/example.go]: I'm referring to the operations in the test file which conflict if run at the same time, which I'd been calling `readVulnerableField`/`writeVulnerableField` above

File: tenets/codelingo/go/writes-to-async-pointer/expected.json [hunk 8]
```diff
@@ -0,0 +1,14 @@
+[
+  {
+   "Comment": "Writing to a reference passed to an asynchronous function can lead to unexpected behaviour.",
```
Comment (fwereade) [inline tenets/codelingo/go/writes-to-async-pointer/expected.json]: is it possible to include a reference to (one of) the conflicting read(s)?

File: tenets/codelingo/go/writes-to-async-pointer/example.go [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+
+type thing struct {
+	name string
+}
+
+func readFromPointer(t *thing) {
+	fmt.Println(t.name)
+}
+
+func safeRead(t thing) {
```
Comment (fwereade) [inline tenets/codelingo/go/writes-to-async-pointer/example.go]: maybe `readFromCopy`?

File: tenets/codelingo/go/writes-to-async-pointer/example.go [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+		t.name = "bus"
+	}(t)
+}
+
+// Concurrent reads and writes to a field of a pointer to an instance of thing. Issue
+func concurrentPointerReadWrites() {
+
+	t := &thing{"Pointer"}
+
+	readVulnerableField(t)
```
Comment (fwereade) [inline tenets/codelingo/go/writes-to-async-pointer/example.go]: similarly, this should never contribute to one of these issues – it's guaranteed to complete before any of the goroutines kick off

File: tenets/codelingo/go/writes-to-async-pointer/codelingo.yaml [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+                    go.names:
+                      go.ident:
+                        name as pointer
+                    go.star_expr
+                  go.assign_stmt(depth = any):
+                    go.lhs(depth = any):
+                      go.ident(depth = any):
+                        name == pointer
+            go.call_expr(depth = any):
+              go.func_lit:
```
Comment (fwereade) [inline tenets/codelingo/go/writes-to-async-pointer/codelingo.yaml]: this is not a comment on this code so much as confirmation we need more sophisticated callgraphs – or perhaps something else

to paraphrase the ideal shape I have in my mind, it's more or less:

    some_memory
    goroutine:
      call_some_function(passing=reference_to(some_memory)):
        access(some_memory)
    anything_else_leading_us_to:
      access(some_memory)

…but not actually triggering when _all_ the accesses are reads

File: tenets/codelingo/go/writes-to-async-pointer/example.go [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+
+	t := &thing{"Pointer"}
+
+	writeVulnerableField(t)
+
+	go writeVulnerableField(t)
+
+	go func(t *thing) {
+		t.name = "bus"
+	}(t)
```
Comment (fwereade) [inline tenets/codelingo/go/writes-to-async-pointer/example.go]: also, would this trigger if we just had

    t := &thing{"x"}
    go t.readFromPointer()
    t.name = "y"

or

    t := &thing{"x"}
    go t.writeToPointer("y")
    fmt.Println(t.name)

?

**Pull request merged**